File: src/engine/entities/PowerUp.java
package engine.entities;


import engine.enums.ObjectType;
import engine.enums.PowerUpType;
import javafx.geometry.Point2D;
import javafx.scene.shape.Circle;
import javafx.scene.shape.Shape;

import java.util.Random;

public class PowerUp extends PhysicsObject {

    private static int nextId = 0;
    /**
     * The unique identifier of this powerup
     */
    private int id;
    /**
     * The type of the power heal,damage or speed
     */
    private PowerUpType type;
    /**
     * If the powerup can be picked up or not
     */
    private boolean isActive = true;

    /**
     * Constructor
     */
    public PowerUp() {
        id = nextId++;
        tag = ObjectType.POWERUP;
        width = tag.getSize();

        //Randomise the type of power up when it spawns
        Random rand = new Random();
        int typeInt = rand.nextInt(3);

        switch (typeInt) {
            case 0:
                type = PowerUpType.HEAL;
                break;
            case 1:
                type = PowerUpType.SPEED;
                break;
            case 2:
                type = PowerUpType.DAMAGE;
                break;
            default:
                break;
        }


        //TODO: Change Values to be based on the map
        int randX = rand.nextInt(2000);
        int randY = rand.nextInt(2000);

        location = new Point2D(randX, randY);
    }

    /**
     * Constructor
     *
     * @param id the id of the power up
     * @param x  x coordinate of the power up
     * @param y  y coordinate of the power up
     */
    public PowerUp(int id, double x, double y) {
        this.id = id;
        this.location = new Point2D(x, y);
    }

    public int getId() {
        return this.id;
    }

    public Point2D getLocation() {
        return this.location;
    }

    public void setLocation(Point2D location) {
        this.location = location;
    }

    /**
     * Activate the power and cause an effect on the player
     *
     * @param player The player you want to be effected by the powerup
     */
    public void activatePowerUp(Character player) {
        if (isActive) {
            switch (type) {
                case HEAL:
                    player.addHealth(10);
                    break;
                case SPEED:
                    player.speedBoost();
                    break;
                case DAMAGE:
                    player.damageBoost();
                    break;
                default:
                    break;
            }
            isActive = false;
        }
    }


    public PowerUpType getType() {
        return type;
    }


    @Override
    public void update() {

    }

    @Override
    public Shape getBounds() {
        return new Circle(location.getX() + width, location.getY() + width, width);
    }
}

File: neon-element-cpp/src/engine/entities/PowerUp.cpp
#include "PowerUp.hpp"
#include "Character.hpp"
#include <random>

namespace neon {

int PowerUp::nextId_ = 0;

PowerUp::PowerUp()
    : id_(nextId_++)
{
    tag_ = ObjectType::PowerUp;
    width_ = objectSize(ObjectType::PowerUp);
    height_ = width_;

    // Randomize type
    static std::mt19937 rng{std::random_device{}()};
    std::uniform_int_distribution<int> typeDist(0, 2);
    switch (typeDist(rng)) {
        case 0: type_ = PowerUpType::Heal; break;
        case 1: type_ = PowerUpType::Speed; break;
        case 2: type_ = PowerUpType::Damage; break;
    }

    // Random position on 2000x2000 map
    std::uniform_int_distribution<int> posDist(0, 2000);
    location_ = {static_cast<float>(posDist(rng)), static_cast<float>(posDist(rng))};
}

PowerUp::PowerUp(int id, float x, float y, PowerUpType type)
    : id_(id), type_(type)
{
    tag_ = ObjectType::PowerUp;
    width_ = objectSize(ObjectType::PowerUp);
    height_ = width_;
    location_ = {x, y};
}

void PowerUp::activatePowerUp(Character& player, const GameClock& clock)
{
    if (!isActive_)
        return;

    switch (type_) {
        case PowerUpType::Heal:
            player.addHealth(10);
            break;
        case PowerUpType::Speed:
            player.speedBoost(clock);
            break;
        case PowerUpType::Damage:
            player.damageBoost(clock);
            break;
    }
    isActive_ = false;
}

} // namespace neon
